modules/icgc-argo-workflows/prep_sample/resources/usr/bin/main.py

```diff
@@ -0,0 +1,355 @@
+
+def filename_to_file(filenames: tuple, files: list) -> tuple:
+    name_to_file_map = {}
+    for f in files:
+      name_to_file_map[os.path.basename(f)] = f
+
+    return (name_to_file_map[filenames[0]], name_to_file_map[filenames[1]] if filenames[1] else None)
+
+def readgroup_id_to_fname(rg_id, input_bam_name='', study_id=None, donor_id=None, sample_id=None):
+    friendly_rgid = "".join([ c if re.match(r"[a-zA-Z0-9\.\-_]", c) else "_" for c in rg_id ])
+    # use original bam file name and rg_id to calculate the md5sum to avoid new lane bam file name collision
+    # use white space (' ') to separate bam name and rg_id
+    md5sum = hashlib.md5(("%s %s" % (input_bam_name, rg_id)).encode('utf-8')).hexdigest()
+
+    if not sample_id or not donor_id or not study_id:
+        sys.exit('Error: missing study/donor/sample ID in the provided metadata')
+
+    return ".".join([study_id, donor_id, sample_id, friendly_rgid, md5sum])
+
+
+def generate_fastqs_from_bam(bam, readgroups, cpu=None, sample_sheet=dict(), study_id=None, donor_id=None, sample_id=None, temp_dir=None, out_dir=None, specimen_id=None, specimen_type=None, tumour_normal_designation=None):
+    # get input bam basename, remove extention to use as output subfolder name
+    bam_base = os.path.splitext(os.path.basename(bam))[0]
+    out_format = bam_base+'/%!.bam'
+    if os.path.exists(bam_base) and os.path.isdir(bam_base):
+      shutil.rmtree(bam_base)
+    os.mkdir(bam_base)
+    cmd = ['samtools', 'split', '-f', '%s' % out_format, '-@ %s' % str(cpu), bam]
+
+    stdout, stderr, returncode = run_cmd(" ".join(cmd))
+    if returncode:
+        sys.exit(f"Error: 'samtools split' failed.\nStdout: {stdout}\nStderr: {stderr}\n")
+
+    # convert readGroupId to filename friendly
+    # only process the lanes output for given input bam
+    for lane_bam in glob.glob(os.path.join(os.getcwd(), bam_base, "*.bam")):
+
+        # remove file extension to get rg_id
+        rg_id = os.path.splitext(os.path.basename(lane_bam))[0]
+
+        # let's make sure RG_ID in lane bam exists in readgroup metadata, either matching read_group_id_in_bam or submitter_read_group_id
+        # otherwise, it should be a lane that's expected to be ignored
+        rg_id_found = False
+        for rg in readgroups:
+            if rg.get('file_r1') == os.path.basename(bam) and (rg.get('read_group_id_in_bam') == rg_id or
+                    (not rg.get('read_group_id_in_bam') and rg['submitter_read_group_id'] == rg_id)):
+                rg_id_found = True
+                # rgs_with_lane_bam_produced.add(rg['submitter_read_group_id'])
+                break
+
+        if rg_id_found:
+            rg_id_fn = readgroup_id_to_fname(rg_id, os.path.basename(bam), study_id, donor_id, sample_id)
+            if rg['is_paired_end']:
+              cmd = f"samtools collate -uO --threads {cpu} {lane_bam} | \
+                samtools fastq -N -O -F 0x900 --threads {cpu} -0 {out_dir}/{rg_id_fn}_other.fq.gz \
```

Comment:
Could you add a comment that `0x900` == `2304` == `not primary alignment + supplementary alignments`

Reply:
Sure. good idea.